controller/src/main/java/io/pravega/controller/server/bucket/PeriodicTxnGC.java

```diff
@@ -0,0 +1,79 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.controller.server.bucket;
+
+import io.pravega.client.stream.Stream;
+import io.pravega.common.hash.RandomFactory;
+import io.pravega.common.tracing.RequestTracker;
+import io.pravega.common.tracing.TagLogger;
+import io.pravega.controller.store.stream.OperationContext;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.task.Stream.StreamTransactionMetadataTasks;
+import io.pravega.controller.util.RetryHelper;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Supplier;
+
+public class PeriodicTxnGC {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(PeriodicTxnGC.class));
+
+    private final StreamTransactionMetadataTasks streamTxnMetadataTasks;
+    private final StreamMetadataStore streamMetadataStore;
+    private final ScheduledExecutorService executor;
+    private final RequestTracker requestTracker;
+    private final Supplier<Long> requestIdGenerator = RandomFactory.create()::nextLong;
+
+    public PeriodicTxnGC(StreamMetadataStore streamMetadataStore, StreamTransactionMetadataTasks streamTxnMetadataTasks,
+                         ScheduledExecutorService executor,
+                         RequestTracker requestTracker) {
+        this.streamTxnMetadataTasks = streamTxnMetadataTasks;
+        this.streamMetadataStore = streamMetadataStore;
+        this.executor = executor;
+        this.requestTracker = requestTracker;
+    }
+    
+    public CompletableFuture<Void> txnGC(Stream stream) {
+        // Track the new request for this automatic transaction GC.
+        long requestId = requestIdGenerator.get();
+        String requestDescriptor = RequestTracker.buildRequestDescriptor("truncateStream", stream.getScope(),
+                stream.getStreamName());
+        requestTracker.trackRequest(requestDescriptor, requestId);
+        OperationContext context = streamMetadataStore.createStreamContext(stream.getScope(), stream.getStreamName(),
+                requestId);
+
+        log.info(requestId, "Periodic background processing for transaction GC called for stream {}/{}", stream.getScope(), stream.getStreamName());
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getOpenTxns(stream.getScope(),
+                stream.getStreamName(), context, executor).thenCompose(openTxnsData -> {
```

Comment:
These should be streamed, IE iterated over rather than all loaded at once. Looking at the underlying implementation, it looks like it already is doing that internally but it's being collected into a list for the sake of the interface.

Reply:
Ok. Sure.